lib/gds_api/test_helpers/email_alert_api.rb

```diff
@@ -68,6 +68,54 @@ def stub_email_alert_api_has_subscription(
           .to_return(status: 200, body: response)
       end
 
+      # Stubs the API responses as if each subscription happened in the order they are passed.
+      # Useful if you need to query the '/latest' endpoint of a subscription.
+      # Takes an array of hashes.
+      #
+      # @example
+      #  stub_email_alert_api_has_subscriptions([
+      #    {
+      #      id: 'id-of-my-subscriber-list',
+      #      frequency: 'weekly',
+      #      ended: true,
+      #    },
+      #    {
+      #      id: 'id-of-my-subscriber-list',
+      #      frequency: 'daily',
+      #    },
+      #  ])
+      #
+      # @param subscriptions [Array]
+      def stub_email_alert_api_has_subscriptions(subscriptions)
+        subscriptions.map! { |subscription| apply_subscription_defaults(subscription) }
+        subscriptions.each do |id, params|
+          latest_id, latest_params = get_latest_matching(params, subscriptions)
+          stub_request(:get, "#{EMAIL_ALERT_API_ENDPOINT}/subscriptions/#{id}")
+            .to_return(status: 200, body: get_subscription_response(id, params).to_json)
+          stub_request(:get, "#{EMAIL_ALERT_API_ENDPOINT}/subscriptions/#{id}/latest")
+            .to_return(status: 200, body: get_subscription_response(latest_id, latest_params).to_json)
+        end
+      end
+
+      def apply_subscription_defaults(subscription)
+        parameters = {
+          title: "Some title",
+          subscriber_id: 1,
+          subscriber_list_id: 1000,
+          ended: true,
+        }.merge(subscription)
+        id = parameters.delete(:id) # Otherwise `ArgumentError: unknown keyword: id`
+        [id, parameters]
+      end
+
+      def get_latest_matching(params, subscriptions)
+        matching = subscriptions.select do |_current_id, current_params|
+          params[:subscriber_id] == current_params[:subscriber_id] &&
+            params[:subscriber_list_id] == current_params[:subscriber_list_id]
+        end
+        matching.last
```

Comment:
Do you need to specify the ordering of results for the `select` query so `.last` is always the final result?  I can't see a `default_scope` for the Subscription model, so the return of results may not always be in a consistent order.

Reply:
> The return of results may not always be in time order.

What do you mean by this? Does Ruby not guarantee the order of arrays being traversed? I know in JavaScript you can't rely on the order of JSON properties being traversed, but don't expect that to be a problem here...?

The `stub_email_alert_api_has_subscriptions` works on the basis that the subscriptions are applied in the order they are passed, i.e. you should not be putting the most recent subscription halfway through the arguments.

Reply:
I see, so the input to `stub_email_alert_api_has_subscriptions` will always be in the correct order, then `.last` is fine here.